File: js/surface.js
/* Script for creating a surface plot render.*/
import { gl, Attribute, 
    TrianglesFrame, makeProgramFromSources } from "./gl-wrappers.js";
import { getShader } from "./shaders.js";

function getVerticesElements(width, height) {
    let elements = [];
    let vertices = [];
    for (let i = 0; i < height; i++) {
        for (let j = 0; j < width; j++) {
            let vertex = [(i + 0.5)/width, (j + 0.5)/height];
            vertex.forEach(e => vertices.push(e));
            if (i < (width - 1) && j < (height - 1)) {
                let triangle1 = [
                    j + i*width, j + 1 + i*width, j + 1 + (i + 1)*width];
                let triangle2 = [
                    j + 1 + (i + 1)*width, j + (i + 1)*width, j + i*width];
                triangle1.forEach(e => elements.push(e));
                triangle2.forEach(e => elements.push(e));
            }
        }
    }
    return [new Float32Array(vertices), 
            (gl.version === 2)? 
            new Int32Array(elements): new Uint16Array(elements)];
}

export default function makeSurface(width, height) {
    let [vertices, elements] = getVerticesElements(width, height);
    return new TrianglesFrame({"position": new Attribute(2, gl.FLOAT, false)},
                              vertices, elements);
}

export function makeSurfaceProgram(fragmentShaderSource) {
    return makeProgramFromSources(getShader("./shaders/surface/vert.vert"), 
                                  fragmentShaderSource);
}

File: c++/dirac_2d/surface.cpp
#include "surface.hpp"

static std::vector<float> get_vertices_set_elements(
    IVec2 d_2d, std::vector<int> &elements) {
    int width = d_2d[0], height = d_2d[1];
    std::vector<float> vertices {};
    for (int i = 0; i < height; i++) {
        for (int j = 0; j < width; j++) {
            std::vector<float> vertex {};
            vertex.push_back(((float)i + 0.5)/float(width));
            vertex.push_back(((float)j + 0.5)/float(height));
            for (float e: vertex)
                vertices.push_back(e);
            if (i < (width - 1) && j < (height - 1)) {
                std::vector<int> triangle1 {
                    j + i*width, j + 1 + i*width, j + 1 + (i + 1)*width
                };
                std::vector<int> triangle2 {
                    j + 1 + (i + 1)*width, j + (i + 1)*width, j + i*width
                };
                for (int e: triangle1)
                    elements.push_back(e);
                for (int e: triangle2)
                    elements.push_back(e);
            }
        }
    }
    return vertices;
}

WireFrame get_surface_wireframe(IVec2 d_2d) {
    Attributes attributes = {
        {"position", {
            .size=2, .type=GL_FLOAT, .normalized=false, .stride=0, .offset=0
        }}
    };
    std::vector<int> elements {};
    std::vector<float> vertices = get_vertices_set_elements(d_2d, elements);
    return WireFrame(attributes, vertices, elements, WireFrame::TRIANGLES);
}